src/libstd/sys/unix/fs.rs

```diff
@@ -51,67 +51,71 @@ pub use crate::sys_common::fs::remove_dir_all;
 
 pub struct File(FileDesc);
 
-// FIXME: This should be available on Linux with all `target_arch` and `target_env`.
-// https://github.com/rust-lang/libc/issues/1545
-macro_rules! cfg_has_statx {
-    ({ $($then_tt:tt)* } else { $($else_tt:tt)* }) => {
-        cfg_if::cfg_if! {
-            if #[cfg(all(target_os = "linux", target_env = "gnu", any(
-                target_arch = "x86",
-                target_arch = "arm",
-                // target_arch = "mips",
-                target_arch = "powerpc",
-                target_arch = "x86_64",
-                // target_arch = "aarch64",
-                target_arch = "powerpc64",
-                // target_arch = "mips64",
-                // target_arch = "s390x",
-                target_arch = "sparc64",
-            )))] {
-                $($then_tt)*
-            } else {
-                $($else_tt)*
-            }
-        }
-    };
-    ($($block_inner:tt)*) => {
-        #[cfg(all(target_os = "linux", target_env = "gnu", any(
-            target_arch = "x86",
-            target_arch = "arm",
-            // target_arch = "mips",
-            target_arch = "powerpc",
-            target_arch = "x86_64",
-            // target_arch = "aarch64",
-            target_arch = "powerpc64",
-            // target_arch = "mips64",
-            // target_arch = "s390x",
-            target_arch = "sparc64",
-        )))]
-        {
-            $($block_inner)*
-        }
-    };
+#[derive(Clone)]
+pub struct FileAttr {
+    stat: stat64,
+    #[cfg(target_os = "linux")]
+    statx_extra_fields: Option<statx::StatxExtraFields>,
 }
 
-cfg_has_statx! {{
+#[cfg(target_os = "linux")]
+mod statx {
+    use super::{cvt, FileAttr};
+    use crate::{io, mem, ptr};
+    use libc::{c_char, c_int, c_uint, off64_t, stat64};
+
+    /// Save extra information from statx.
     #[derive(Clone)]
-    pub struct FileAttr {
-        stat: stat64,
-        statx_extra_fields: Option<StatxExtraFields>,
+    pub(super) struct StatxExtraFields {
+        // This is needed to check if btime is supported by the filesystem.
+        pub stx_mask: u32,
+        pub stx_btime_tv_sec: i64,
+        pub stx_btime_tv_nsec: u32,
+    }
+
+    // We need to port these structs here, since libc does not always contain
+    // them, eg. on musl.
+    #[repr(C)]
+    struct statx {
```

Comment:
This is scary. libc has CI in place to verify that we don’t end up with wrong struct definitions, but that’s not true for rustc… I think it makes more sense overall to just not use statx on the environments that do not want to expose this type.

Reply:
Yes this should be using the definitions in `libc` rather than duplicating them here.

Reply:
But musl does not export these structs currently... So we should only enable this when using glibc on Linux?

Reply:
Musl does support `statx` since 1.1.24: https://gitlab.com/gsauthof/musl/commit/dfc81828f7ab41da08f744c44117a1bb20a05749

Reply:
@mati865 No yet. It never exposes `statx` function wrapper and related structs. It just use it internally.
There's no way to get file birth time from public API, except invoking `syscall` with manual defined `struct statx`.

Reply:
To land, this *must* be using the `statx` structure from a location that is verified to be correct. The `libc` crate is one source which is verified correct, but it doesn't have to be the only one. Linux structures are notorious for changing across architectures and across libraries, so having an unverified copy here won't be maintainable over time.